File: node-express/src/routes/note.ts
import { 
    randomUUID 
} from "crypto";

//-----------------------------------------------------------
//														Models

type Note = {
	title: string;
	content: string;
	createdAt: Date
}

type UserNote = {
	notes:Record<string, Note>
}

//-----------------------------------------------------------
//														Declarations

const userNotes:Record<string, UserNote> = {};

export const createNote = async (req:any, res:any) => {
    // TODO: fake auth
    const {userid} = req.headers as any;
    if(!userid){
        return res.status(401).send("Invalid properties");
    }

    const {title, content} = req.body as any;
    if(!title || !content){
        return res.status(400).send("Invalid properties");
    }

    let userNote = userNotes[userid];
    if(!userNote){
        userNote = {
            notes: {}
        };
        userNotes[userid] = userNote;
    }

    const noteId = randomUUID();
    userNote.notes[noteId] = {
        title,
        content,
        createdAt: new Date()
    };

    res.status(201).json({ "id": noteId });
}

export const getNotes = async (req:any, res:any) => {
    // TODO: fake auth
    const {userid} = req.headers as any;
    if(!userid){
        return res.status(401).send("Invalid properties");
    }

    const userNote = userNotes[userid];
    if(!userNote){
        return res.json({"items":[]});
    }

    res.json({"items":userNote.notes});
}

export const getNoteById = async (req:any, res:any) => {
    // TODO: fake auth
    const {userid} = req.headers as any;
    if(!userid){
        return res.status(401).send("Invalid properties");
    }

    const { id } = req.params as any;

    const userNote = userNotes[userid];
    if(!userNote){
        return res.status(404).send("");
    }

    const note = userNote.notes[id];
    if(!note){
        return res.status(404).send("");
    }

    res.json({"item": note});
}

File: rust-actix/src/routes/note.rs
use std::{sync::Mutex, collections::HashMap};

use actix_web::{web, HttpResponse, HttpRequest};
use serde::{Serialize, Deserialize};
use serde_json::json;
use uuid::Uuid;

use crate::models::model::{NoteData, Note, UserNoteData};


#[derive(Debug, Deserialize)]
pub struct InputNote {
    pub title: String,
    pub content: String,
}

#[derive(Serialize, Clone)]
pub struct NoteOutput {
    pub items: HashMap<String, Note>,
}

#[derive(Deserialize)]
pub struct NoteIdInput {
    id: String,
}

fn get_userid_header<'a>(req: &'a HttpRequest) -> Option<&'a str> {
    req.headers().get("userid")?.to_str().ok()
}


pub async fn create_note(users: web::Data<Mutex<UserNoteData>>, req: HttpRequest, item: web::Json<InputNote>)-> HttpResponse {
    let user_id_header = get_userid_header(&req);
    if user_id_header.is_none() {
        return HttpResponse::Unauthorized().body("");
    }

    let user_id = user_id_header.unwrap();

    let mut data = users.lock().unwrap();
    let note_data_option = data.users.get(user_id);
    

    let id = Uuid::new_v4().to_string();
    let note = Note {
        title: item.title.clone(),
        content: item.content.clone(),
    };

    if note_data_option.is_none() {
        let notes = Mutex::new(HashMap::new());
        notes.lock().unwrap().insert(id.clone(), note);

        let note_data = NoteData{
            notes: notes
        };

        data.users.insert(user_id.to_string(), note_data);
    } else {
        note_data_option.unwrap().notes.lock().unwrap().insert(id.clone(), note);
    }

    let result = json!({
        "id": id,
    });

    HttpResponse::Created().json(web::Json(result))
}

pub async fn  get_notes(users: web::Data<Mutex<UserNoteData>>, req: HttpRequest)-> HttpResponse {
    let user_id_header = get_userid_header(&req);
    if user_id_header.is_none() {
        return HttpResponse::Unauthorized().body("");
    }
    let user_id = user_id_header.unwrap();

    let data = users.lock().unwrap();
    let note_data = data.users.get(user_id);

    let result:NoteOutput;
    if note_data.is_none() {
        result = NoteOutput {
            items: HashMap::new()
        };
    }else {
        result = NoteOutput {
            items: note_data.unwrap().notes.lock().unwrap().clone()
        };
    }
    
    HttpResponse::Ok().json(result)
}

pub async fn get_note_by_id(users: web::Data<Mutex<UserNoteData>>, req: HttpRequest, note_input: web::Path<NoteIdInput>,)-> HttpResponse {
    let user_id_header = get_userid_header(&req);
    if user_id_header.is_none() {
        return HttpResponse::Unauthorized().body("");
    }
    let user_id = user_id_header.unwrap();

    let data = users.lock().unwrap();
    let note_data = data.users.get(user_id);
    if note_data.is_none() {
        return HttpResponse::NotFound().body("");
    }
    

    let note_data = note_data.unwrap();
    let notes = note_data.notes.lock().unwrap();
    let note = notes.get(&note_input.id);
    
    if note.is_none() {
        return HttpResponse::NotFound().body("");
    }
    
    let result = json!({
        "item": note, 
    });

    HttpResponse::Ok().json(result)
}